.github/workflows/main.yml

```diff
@@ -8,6 +8,11 @@ jobs:
     steps:
       - uses: actions/checkout@v4
       - uses: dtolnay/rust-toolchain@stable
+      - name: Install dependencies on ubuntu
+        run: |
+          sudo apt-get update
+          sudo apt-get install llvm libelf-dev clang
+          clang --version
```

Comment:
i'm going to say this is going to be an issue for us at some point - most of our users won't have llvm/clang installed so they'll all run into cryptic build errors. We need some user-friendly "Warning, you need these packages installed" during the cargo build process.

And/or a github artifact for direct download (filed as issue #29 so we don't forget)

Reply:
One solution could be what is done in the kernel for pre-loaded bpf: we generate the skeleton and then check in that output. Then we add a flag to cargo to use the embedded generated skeleton or regenerate it (with potentially some way to detect if they are out of sync, maybe with `git diff`).

> And/or a github artifact for direct download (filed as issue https://github.com/hidutils/hid-recorder/issues/29 so we don't forget)

Thanks for that!

Reply:
I've managed to turn `clang` optional by adding the generated `.skel.rs` to the source tree and calling or not `SkeletonBuilder`.

However, I can not make `libelf-dev` (or `elfutils-libelf-devel` on fedora) optional: when libppf-rs is pulled, it pulls libbpf-sys which compiles itself and fails at `libelf.h` not found.

 I could hide the bpf functionality behind a feature, which would make `libbpf-rs` optional, but then we would need to figure out whether we want it enabled by default or not. And in the end, we probably just ask users to download the binary from #29, so not sure it's worth the effort.